lib/Dialect/AIG/AIGOps.cpp

```diff
@@ -0,0 +1,141 @@
+//===- LoopScheduleOps.cpp - LoopSchedule CIRCT Operations ------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implement the AIG ops.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/AIG/AIGOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "mlir/IR/PatternMatch.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace circt::aig;
+
+#define GET_OP_CLASSES
+#include "circt/Dialect/AIG/AIG.cpp.inc"
+
+OpFoldResult AndInverterOp::fold(FoldAdaptor adaptor) {
+  // TODO: Implement this.
+  return {};
+}
+
+LogicalResult AndInverterOp::canonicalize(AndInverterOp op,
+                                          PatternRewriter &rewriter) {
+  // TODO: Extend to more than 2 inputs.
+  if (op.getInputs().size() == 2) {
+    if ((op.getInputs()[0] == op.getInputs()[1]) &&
+        (op.isInverted(0) == op.isInverted(1))) {
+      auto all = rewriter.create<hw::ConstantOp>(
+          op.getLoc(), APInt::getAllOnes(op.getType().getIntOrFloatBitWidth()));
+      rewriter.replaceOpWithNewOp<aig::AndInverterOp>(op, op.getInputs()[0],
+                                                      all, true, false);
+
+      return success();
+    }
+  }
+  return failure();
+}
+
+mlir::ParseResult AndInverterOp::parse(mlir::OpAsmParser &parser,
+                                       mlir::OperationState &result) {
+  SmallVector<OpAsmParser::UnresolvedOperand> operands;
+  SmallVector<bool> inverts;
+  auto loc = parser.getCurrentLocation();
+
+  while (true) {
+    if (succeeded(parser.parseOptionalKeyword("not"))) {
+      inverts.push_back(true);
+    } else {
+      inverts.push_back(false);
+    }
+    operands.push_back(OpAsmParser::UnresolvedOperand());
```

Comment:
Could this be done with something like the custom bool keyword parser/printer that we have somewhere in `circt/Support`?

Reply:
We need to modify parseKeywordBool quite a bit since it currently takes BoolAttr& and both trueKeyword/falseKeyword so I think it's fine to keep separated implementation.

Reply:
That makes total sense!